### PR TITLE
Enable upsample2x optimization for opset 11 Resize

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
Add ability to enable upsample2x optimization for Resize in opset 11 by adding checks on the various attributes and inputs.
Do as much of the check upfront as possible. Do at runtime if necessary.

**Motivation and Context**
Address #3297 

With model from 3297 (Windows RelWithDebInfo build of ORT)
Before: Average inference time cost:26.1289 ms
After: Average inference time cost:0.407622 ms

### CODE REVIEW

File: onnxruntime/core/providers/cpu/tensor/upsample.cc [hunk 1]
```diff
@@ -581,6 +583,119 @@ void ResizeBiCubic(
...
+  // check scales
+  // Cast the scales to integers and verify that the scales are positive and
+  // round trip back to floating point.
+  std::vector<int64_t> scales_attr(4);
+  for (size_t n = 0; n < 4; n++) {
+    int64_t scale_value = static_cast<int64_t>(scales[n]);
+    if (scale_value <= 0 || static_cast<float>(scale_value) != scales[n]) {
+      return false;
+    }
+    scales_attr[n] = scale_value;
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/providers/cpu/tensor/upsample.cc:688]: Should this also check the last 2 dims are 2? The only place use_nearest2x_optimization appears to be used requires that to be the case, so I wasn't sure if other values such as {1,1,4,4} were acceptable or not. 

    if (use_nearest2x_optimization && scales[0] == 1 && scales[1] == 1 && scales[2] == 2 && scales[3] == 2) {


---
In reply to: [401873838](https://github.com/microsoft/onnxruntime/pull/3388#discussion_r401873838) [](ancestors = 401873838)

File: onnxruntime/core/providers/cpu/tensor/upsample.cc [hunk 2]
```diff
@@ -581,6 +583,119 @@ void ResizeBiCubic(
...
+      if (sizes_shape != nullptr && utils::HasDimValue(sizes_shape->dim()[0]) &&
+          sizes_shape->dim()[0].dim_value() != 4) {
+        return;  // optimization not possible
+      }
+
+      if (!x_shape || !sizes_shape) {
+        break;  // need to check at runtime
+      }
+
+      if (!sizes_shape->dim()[0].has_dim_value()) {
```
[the remark this answers is not in the document]
Comment (skottmckay) [reply onnxruntime/core/providers/cpu/tensor/upsample.cc:637]: It should be HasDimValue. I initially started with it folded in but it made that condition fairly ugly, so for readability I separated it out (this code is called during kernel creation so it's one-off).

---
In reply to: [401865755](https://github.com/microsoft/onnxruntime/pull/3388#discussion_r401865755) [](ancestors = 401865755)

File: onnxruntime/core/providers/cpu/tensor/upsample.h [hunk 3]
```diff
@@ -129,6 +131,7 @@ class UpsampleBase {
   float cubic_coeff_a_;
   bool exclude_outside_;
   float extrapolation_value_;
+  bool can_use_nearest2x_optimization_with_scales_check_ = false;
```
Comment (tracysh) [inline onnxruntime/core/providers/cpu/tensor/upsample.h:134]: no longer needed?

File: onnxruntime/core/providers/cpu/tensor/upsample.cc [hunk 4]
```diff
@@ -3,6 +3,8 @@
 
 #include "core/providers/cpu/tensor/upsample.h"
 #include "core/common/safeint.h"
+#include "core/graph/graph_utils.h"
```
Comment (tracysh) [inline onnxruntime/core/providers/cpu/tensor/upsample.cc:6]: are these includes still needed?

File: onnxruntime/core/providers/cpu/tensor/upsample.h (hunk 3 above)
Comment (skottmckay) [reply onnxruntime/core/providers/cpu/tensor/upsample.h]: Sorry - overlooked that.

**Pull request merged**